This directory contains all the scripts for producing and testing a mountain gorilla target SNP panel.  

- Part 1: Identifying target loci - [marker_design](https://github.com/sudmantlab/rishi/tree/main/gorilla_census/marker_design)
  - [00_data_download.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/marker_design/00_data_download.md) includes information on previously published data download (from SRA)
  - [01_mapping.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/marker_design/01_mapping.md) explains mapping these data and filtering the resulting bam files
  - [02_genotyping.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/marker_design/02_genotyping.md) covers genotyping and vcf file filtering (for quality etc.)
  - [03_PCA.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/marker_design/03_PCA.md) explains the production of PCAs from this data
  - [04_geno.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/marker_design/04_geno.md) outlines how to convert the vcf files to .geno files and then carry out popgen analyses on these files
  - [05_mpcrselect](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/marker_design/05_mpcrselect.md) covers running the mPCRselect pipeline to select putative target SNPs
  - [06_snp_panel_popgen.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/marker_design/06_snp_panel_popgen.md) plotting PCA from mpcrselect vcf output
  - [07_primer_set_selection.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/marker_design/07_primer_set_selection.md) get #SNPs and plot PCAs for each ~120-150 snp set to select one and outlines selected set
  - *Summary: The result of this work was that we produced a VCF containing high diversity SNPs (across the set of gorilla data previously published) that were then used by the Petrov lab to produce a set of compatible 2 and then 1 step PCR primers.*
- Part 2: Initial 2-step PCR test - [Jan_2step_prc_test](https://github.com/sudmantlab/rishi/tree/main/gorilla_census/Jan_2step_prc_test)
  - [01_fastp.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/Jan_2step_prc_test/01_fastp.md) includes QC of reads using fastp
  - [02_mapping.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/Jan_2step_prc_test/02_mapping.md) outlines mapping commands
  - [03_genotyping.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/Jan_2step_prc_test/03_genotyping.md) outlines genotyping of loci for test
  - [04_PCA.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/Jan_2step_prc_test/04_PCA.md) code to make pcas from test loci
  - [05_refine_loci.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/Jan_2step_prc_test/05_refine_loci.md) covers how to identify and export possible new high-pi target SNPs for a refined second test
  - *Summary: The result of this test sequencing of 156 SNPs was that we were able to amplify and genotype the target loci but there were some concerns about the potential for unbarcoded reads amplified from the first PCR step to contaminate the lab/other samples and so testing a 1-step PCR where amplicons are amplified with barcode attached was suggested.*
- Part 3: Follow up 1-step PCR test - [July_1step_pcr_test](https://github.com/sudmantlab/rishi/tree/main/gorilla_census/July_1step_pcr_test)
  - [01_QC_and_demultiplex.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/July_1step_pcr_test/01_QC_and_demultiplex.md) contains code to QC and demultiplex the single fastq file for this test
  - [02_mapping.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/July_1step_pcr_test/02_mapping.md) code to map reads to gorilla genome
  - [03_genotyping.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/July_1step_pcr_test/03_genotyping.md) contains info on genotyping and filtering/processing of genotype information for all 40 target snps
  - [04_plot_genotypes.md](https://github.com/sudmantlab/rishi/blob/main/gorilla_census/July_1step_pcr_test/04_plot_genotypes.md) code for plotting the genotypes in the test in a heatmap
  - *Summary: This test showed that of the 40 SNPs we targeted we were able to genotype, however, consistency of genotype calls across replicates for each sample was a challenge. Additionally, some barcode combinations that were not provided any template (i.e. blanks) produced barcoded reads and were genotypes highlighting the potential for this approach to result in false positive reads perhaps through chimerisation.*
